cppcache/src/PdxHelper.cpp

```diff
@@ -47,92 +47,105 @@ PdxHelper::PdxHelper() {}
+    pdxInstance->setPdxId(typeId);
   }
+  auto pdxLocalWriter =
+      PdxLocalWriter(output, pdxInstance->getPdxType(), pdxTypeRegistry);
+  pdxInstance->toData(pdxLocalWriter);
+  pdxLocalWriter.endObjectWriting();  // now write typeid
+  auto pdxStream = pdxLocalWriter.getPdxStream();
+  pdxInstance->updatePdxStream(pdxStream, pdxLocalWriter.getBytesWritten());
+  delete[] pdxStream;
+}
 
+PdxRemoteWriter PdxHelper::createPdxRemoteWriter(
+    const std::shared_ptr<PdxTypeRegistry>& pdxTypeRegistry,
+    const std::shared_ptr<PdxSerializable>& pdxObject, DataOutput& output,
+    const std::string& className) {
+  if (auto pd = pdxTypeRegistry->getPreserveData(pdxObject)) {
+    auto mergedPdxType = pdxTypeRegistry->getPdxType(pd->getMergedTypeId());
+    return PdxRemoteWriter(output, mergedPdxType, pd, pdxTypeRegistry);
+  } else {
+    return PdxRemoteWriter(output, className, pdxTypeRegistry);
+  }
+}
+
+void PdxHelper::serializePdxSerializable(
+    const std::shared_ptr<PdxSerializable>& pdxObject, CacheImpl* cacheImpl,
+    std::shared_ptr<PdxTypeRegistry>& pdxTypeRegistry, DataOutput& output) {
   auto&& className = pdxObject->getClassName();
-  auto localPdxType = pdxTypeRegistry->getLocalPdxType(className);
-
-  if (localPdxType == nullptr) {
-    // need to grab type info, as fromdata is not called yet
-
-    PdxWriterWithTypeCollector ptc(output, className, pdxTypeRegistry);
-    pdxObject->toData(ptc);
-    auto nType = ptc.getPdxLocalType();
-
-    nType->InitializeType();
-    int32_t nTypeId = pdxTypeRegistry->getPDXIdForType(
-        className.c_str(), DataOutputInternal::getPool(output), nType, true);
-    nType->setTypeId(nTypeId);
-
-    ptc.endObjectWriting();
-    pdxTypeRegistry->addLocalPdxType(className, nType);
-    pdxTypeRegistry->addPdxType(nTypeId, nType);
-
-    if (cacheImpl != nullptr) {
-      uint8_t* stPos = const_cast<uint8_t*>(output.getBuffer()) +
-                       ptc.getStartPositionOffset();
-      int pdxLen = PdxHelper::readInt32(stPos);
-      cachePerfStats.incPdxSerialization(
-          pdxLen + 1 + 2 * 4);  // pdxLen + 93 DSID + len + typeID
-    }
 
-  } else  // we know locasl type, need to see preerved data
-  {
-    // if object got from server than create instance of RemoteWriter otherwise
-    // local writer.
-    // now always remotewriter as we have API Read/WriteUnreadFields
-    // so we don't know whether user has used those or not;; Can we do some
-    // trick here?
-
-    auto createPdxRemoteWriter = [&]() -> PdxRemoteWriter {
-      if (auto pd = pdxTypeRegistry->getPreserveData(pdxObject)) {
-        auto mergedPdxType = pdxTypeRegistry->getPdxType(pd->getMergedTypeId());
-        return PdxRemoteWriter(output, mergedPdxType, pd, pdxTypeRegistry);
-      } else {
-        return PdxRemoteWriter(output, className, pdxTypeRegistry);
-      }
-    };
+  if (!pdxTypeRegistry->getLocalPdxType(className)) {
```

Comment:
This condition statement should be wrapped as a predicate, and getLocalPdxType should throw an exception instead of return null.